### DISCUSSION TITLE
Some questions

### DISCUSSION THREAD

**Discussion by Ra2-IFV** [twpayne/chezmoi#4001] [General]
So I have some questions...

## includeTemplate doen't follow .chezmoiroot

Inside `.chezmoiroot` is `home`, and there is `home/foobar.tmpl`.  
If I use `{{ includeTemplate "foobar.tmpl" }}` in`home/.chezmoi.toml.tmpl`, chezmoi will report:

`executing "chezmoi.toml" at <includeTemplate "foobar.tmpl">: error calling includeTemplate: open ~/.local/share/chezmoi/foobar.tmpl: no such file or directory`
even if I put it in the `home/.chezmoitemplates` directory. I have to use this instead:
`{{ includeTemplate "home/foobar.tmpl" }}`

This line works in other template files. I guess it's a bug?

## {{ $key := value }} doen't work if it's in the {{ if }} {{ end }} directive
This works:
```
{{ $foo := "bar" }}

key = "{{ $foo }}"
```
And this not:
```
{{ if CONDITION }}
{{   $foo := "bar" }}
{{ end }}

key = "{{ $foo }}"
```
chezmoi reports `$foo` is undefined. Is this intentional? I do did a search but there are 200+ issues...

**Comment by halostatue** [comment]
> `{{ $key := value }}` doesn't work if it's in the `{{ if }}` `{{ end }}` directive

This is not a bug in chezmoi, but a misunderstanding of Go scoping rules and semantics.

- Variables are *defined* with `:=`. They are reassigned with `=`.
- Variables may be *redefined* in new scopes. Within Go, scopes are introduced by brackets (`{}`).  In Go *templates*, since braces are used for template markers, there are no brackets and scopes are introduced by specific blocks (like `if`, `range`, etc.; basically structures that require `end`).
- Variables are only available in the scopes where they are defined.

So this will work:

```console
$ chezmoi execute-template '
{{- $key := "baz" -}}
{{- if gt 3 1 -}}
{{-  $key = "bar" -}}
{{- end -}}
{{ $key }}'
bar
```

Note that this will not do what you expect:

```console
$ chezmoi execute-template '
{{- $key := "baz" -}}
{{- if gt 3 1 -}}
{{-  $key := "bar" -}}
{{- end -}}
{{ $key }}'
baz
```

The only difference between the use is `=` vs `:=` inside the conditional.

**Reply by Ra2-IFV** [reply]
So if I use `:=` inside `if end`, it will **define** a new variable, that only exists and usable in the `if` scope. I need to **define** it outside the `if` blocks, then assign new values.

Also some experiments:  if there is `{{ $a = 1 }}` in `a.tmpl`, calling `include` or `includeTemplate` in `b.tmpl` will not make `$a` being defined in `b.tmpl`.

**Reply by twpayne** [reply]
You are asking questions about the Go's [`text/template`](https://pkg.go.dev/text/template) language, not about chezmoi. Please read chezmoi's [templating guide](https://www.chezmoi.io/user-guide/templating/) and [general documentation on Go's `text/template`](https://duckduckgo.com/?q=go+text%2Ftemplate&ia=web).

**Comment by halostatue** [comment]
> > `includeTemplate` doesn't follow `.chezmoiroot`
> Inside `.chezmoiroot` is `home`, and there is `home/foobar.tmpl`. If I use `{{ includeTemplate "foobar.tmpl" }}` in`home/.chezmoi.toml.tmpl`, chezmoi will report:
> 
> `executing "chezmoi.toml" at <includeTemplate "foobar.tmpl">: error calling includeTemplate: open ~/.local/share/chezmoi/foobar.tmpl: no such file or directory` even if I put it in the `home/.chezmoitemplates` directory. I have to use this instead: `{{ includeTemplate "home/foobar.tmpl" }}`
> 
> This line works in other template files. I guess it's a bug?

This might not be a bug (one should probably *not* be using `includeTemplate` in the config template), but I have reported #4002 so that it can be investigated.
